File: python/gunmu/collect.py
from __future__ import annotations

from collections.abc import Iterable, Mapping
from typing import TypeVar

from .core import Gunmu, Nothing, Some

T = TypeVar("T")


def collect(items: Iterable[Gunmu[T]]) -> Gunmu[list[T]]:
    """收集所有 Some；遇到第一个 Nothing 立即返回。"""
    values: list[T] = []

    for item in items:
        if isinstance(item, Nothing):
            return item

        values.append(item.unwrap())

    return Some(values)


def collect_all(items: Iterable[Gunmu[T]]) -> Gunmu[list[T]]:
    """收集所有值；若存在 Nothing，则合并所有 reason。"""
    values: list[T] = []
    reasons: list[str] = []

    for item in items:
        if isinstance(item, Nothing):
            reasons.append(item.reason())
        else:
            values.append(item.unwrap())

    if reasons:
        return Nothing("; ".join(reasons))

    return Some(values)


def collect_dict(
    mapping: Mapping[str, Gunmu[T]],
) -> Gunmu[dict[str, T]]:
    """收集字典中的所有 Some；遇到 Nothing 则失败。"""
    result: dict[str, T] = {}

    for key, item in mapping.items():
        if isinstance(item, Nothing):
            return Nothing(f"{key}: {item.reason()}")

        result[key] = item.unwrap()

    return Some(result)

File: typescript/gunmu/collect.ts
/**
 * 把一组 Gunmu 收成一个 Gunmu<T[]>。
 */

import { Gunmu, some, nothing, isSome } from "./core.js";

export function collect<T>(items: Iterable<Gunmu<T>>): Gunmu<T[]> {
  const values: T[] = [];
  for (const item of items) {
    if (!isSome(item)) {
      return item;
    }
    values.push(item.value);
  }
  return some(values);
}

export function collectAll<T>(items: Iterable<Gunmu<T>>): Gunmu<T[]> {
  const values: T[] = [];
  const reasons: string[] = [];
  for (const item of items) {
    if (isSome(item)) {
      values.push(item.value);
    } else {
      reasons.push(item.reason);
    }
  }
  if (reasons.length > 0) {
    return nothing(reasons.join("; "));
  }
  return some(values);
}

export function collectDict<T>(
  map: Record<string, Gunmu<T>>,
): Gunmu<Record<string, T>> {
  const result: Record<string, T> = {};
  for (const [key, item] of Object.entries(map)) {
    if (isSome(item)) {
      result[key] = item.value;
    } else {
      return nothing(`${key}: ${item.reason}`);
    }
  }
  return some(result);
}